### PR TITLE
Stage class

### LINKED ISSUE #2: Stage class
Manages the map layout, obstacles, walls etc.
- 2D array for layout

### PR DESCRIPTION
**Pull request by chesterhow**
Changes
- Created Stage class to manage 2D array of level

@ivantjh @JxJoker 

Fixes #2

### CODE REVIEW

File: src/stage.h [hunk 1]
```diff
@@ -0,0 +1,39 @@
+#ifndef _STAGE_H
+#define _STAGE_H
+#define WIN32_LEAN_AND_MEAN
+
+#include <iostream>
+#include <string>
+#include <vector>
+#include "constants.h"
+
```
Comment (ivantjh) [inline src/stage.h:10]:
```c++
struct TileCoord {
    int x;
    int y;
}

struct MapCoord {
    float x;
    float y;
}
```

Use this for startTile, endTile, currentTile and all the corresponding methods.

Comment (ivantjh) [reply src/stage.h:10]: Don't think it is necessary to have a vector only storing two values.

File: src/stage.h [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+class Stage {
+private:
+	std::vector<std::vector<int>> layout;	// 2D array representing stage layout
+	std::vector<int> startTile;				// Coordinates of start tile
+	std::vector<int> endTile;				// Coordinates of end tile
+	std::vector<int> currentTile;			// Coordinates of current tile
```
Comment (ivantjh) [inline src/stage.h]: Need to keep `stageNo` variable to know which stage player is on.

File: src/stage.h [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+	Stage();
+
+	// Destructor
+	virtual ~Stage();
+
+	bool initialize(std::vector<std::vector<int>> layout, std::vector<int> startTile, std::vector<int> endTile);
+
+	void logLayout();
+	
+	int getTileAtCoordinate(std::vector<int> coordinate);
```
Comment (ivantjh) [inline src/stage.h]: `getTileValueAtMapCoord(MapCoord mc)` to keep it consistent with the bottom method name

File: src/threechances.cpp [hunk 4]
```diff
@@ -24,6 +24,59 @@ ThreeChances::~ThreeChances() {
...
+	// Legend
+	// 0 - border
+	// 1 - 1st floor path
+	// 2 - 2nd floor path
+	// 3 - stairs
+	// 4 - water
+	// 5 - obstacle
+	// 6 - blockage
+
+	std::vector<std::vector<int>> layout = {
```
Comment (ivantjh) [inline src/threechances.cpp]: Create a new file named 'mapLayout.h' or something relevant and store there. The main game file shouldn't be used to store large data.

**Review by ivantjh — CHANGES_REQUESTED** [review]
Looks okay, not sure if we need setters and getters for each of the private variables in the stage class.

File: src/stage.h (hunk 1 above)
Comment (chesterhow) [reply src/stage.h:10]: Whats the diff between TileCoord and MapCoord?

Comment (ivantjh) [reply src/stage.h:10]: TileCoord is used to store the indexes of the 2D array. MapCoord is used to store the actual x, y coordinates of the map.

Comment (chesterhow) [reply src/stage.h:10]: But i only have coordinates. Which is just the indexes. Having 2 will only screw us over

**Comment by chesterhow** [comment]
@ivantjh @JxJoker made the relevant changes

File: src/stageLayouts.h [hunk 5]
```diff
@@ -0,0 +1,56 @@
+#include <vector>
+
+struct Coordinates {
+	int x;
+	int y;
+	Coordinates() {}
+	Coordinates(int x, int y) : x(x), y(y) {}
```
Comment (ivantjh) [inline src/stageLayouts.h:7]: Need constructors? Or is that constructor?

**Review by ivantjh — APPROVED** [review]

File: src/stageLayouts.h (hunk 5 above)
Comment (chesterhow) [reply src/stageLayouts.h:7]: Dats right

**Pull request merged**